pkg/parquetconverter/parquet_converter_ring_test.go

```diff
@@ -0,0 +1,215 @@
+			now := time.Now()
+			mockTSDB(t, path.Join(storageDir, "user-1"), 24, numBlocks/2, now.Add(-24*time.Hour).Unix()*1000, now.Unix()*1000)
+			mockTSDB(t, path.Join(storageDir, "user-2"), 24, numBlocks/2, now.Add(-24*time.Hour).Unix()*1000, now.Unix()*1000)
+
+			// Uploading the blocks to in memory storage - otherwise we run out of file descriptors during the test
+			// The default limit for fsd is 1024 on linux.
+			bucketClient := objstore.NewInMemBucket()
+			for _, userID := range []string{"user-1", "user-2"} {
+				userBucketClient := bucket.NewUserBucketClient(userID, bucketClient, nil)
+				require.NoError(t, bucketClientOnDisk.Iter(context.Background(), userID, func(key string) error {
+					dir := strings.TrimSuffix(path.Join(storageDir, key), "/")
+					err := block.Upload(context.Background(), log.NewNopLogger(), userBucketClient, dir, nil)
+					if err != nil {
+						return err
+					}
+					return nil
+				}))
+			}
+
+			ringStore, closer := consul.NewInMemoryClientWithConfig(ring.GetCodec(), consul.Config{
+				MaxCasRetries: 20,
+				CasRetryDelay: 500 * time.Millisecond,
+			}, log.NewNopLogger(), nil)
+			t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+			var converters []*ParquetConverter
+			registries := dskit_metrics.NewTenantRegistries(log.NewNopLogger())
+
+			for i := 1; i <= numConverters; i++ {
+				instanceID := fmt.Sprintf("converter-%d", i)
+
+				converterCfg := prepareConfig(t)
+				converterCfg.ShardingRing.Common.KVStore.Mock = ringStore
+				converterCfg.ShardingRing.Common.InstanceID = instanceID
+				converterCfg.ShardingRing.Common.InstanceAddr = fmt.Sprintf("127.0.0.%d", i)
+				converterCfg.ShardingRing.WaitStabilityMinDuration = 0
+				converterCfg.ShardingRing.WaitStabilityMaxDuration = 0
+				converterCfg.ShardingRing.WaitActiveInstanceTimeout = 30 * time.Second
+				converterCfg.ConversionInterval = 1 * time.Second
+
+				c, reg := prepareWithMockConverter(t, converterCfg, objstore.WithNoopInstr(bucketClient))
+				t.Cleanup(func() {
+					err := services.StopAndAwaitTerminated(context.Background(), c)
+					if err != nil && errors.Is(err, context.Canceled) {
+						return
+					}
+					assert.NoError(t, err)
+				})
+
+				converters = append(converters, c)
+				registries.AddTenantRegistry(instanceID, reg)
+			}
+
+			// Start all converters concurrently.
+			for _, g := range converters {
+				require.NoError(t, g.StartAsync(ctx))
+			}
+
+			// Wait until all converters are running.
+			for _, g := range converters {
+				require.NoError(t, g.AwaitRunning(ctx))
+			}
+
+			// At this point we expect that all converters have joined the ring, discovered all tenants, and converted
+			// blocks for all users.
+
+			// All converters should have discovered the same number of tenants.
+			test.Poll(t, 10*time.Second, true, func() interface{} {
+				metrics := registries.BuildMetricFamiliesPerTenant()
+				users := metrics.GetSumOfGauges("cortex_parquet_converter_tenants_discovered")
```

Comment:
Could be saved for a later PR: this number makes more sense in conjunction with the corresponding metric like `cortex_compactor_tenants_skipped` which is incremented when a discovered tenant is skipped because it is not owned by the compactor due to the sharding.

Reply:
Thought about this but the sharding for the converter is per block not per tenant. Would you want a counter with blocks discarded?